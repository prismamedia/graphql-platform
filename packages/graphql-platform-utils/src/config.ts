export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
