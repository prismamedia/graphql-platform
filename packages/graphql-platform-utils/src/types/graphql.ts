export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
