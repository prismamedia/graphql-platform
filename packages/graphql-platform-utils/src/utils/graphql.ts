export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
