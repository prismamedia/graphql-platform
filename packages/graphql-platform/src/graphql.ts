import type * as graphql from 'graphql';
import type { ConnectorInterface } from './connector-interface.js';
import type { GraphQLPlatform } from './index.js';

export type GPBoundGraphQLFieldResolver<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> = (
  this: GraphQLPlatform<TRequestContext, TConnector>,
  source: TSource,
  args: TArgs,
  context: TRequestContext,
  info: graphql.GraphQLResolveInfo,
) => TResult;

export interface GPBoundGraphQLFieldConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> extends graphql.GraphQLFieldConfig<TSource, TRequestContext, TArgs> {
  resolve?: GPBoundGraphQLFieldResolver<
    TRequestContext,
    TConnector,
    TSource,
    TArgs,
    TResult
  >;
}
