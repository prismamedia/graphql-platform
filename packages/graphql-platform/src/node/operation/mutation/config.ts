import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
