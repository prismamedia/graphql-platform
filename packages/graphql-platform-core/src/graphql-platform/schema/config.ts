import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
