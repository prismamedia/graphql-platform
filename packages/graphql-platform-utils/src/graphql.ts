import * as graphql from 'graphql';
import { getOptionalFlag, OptionalFlag } from './config.js';
import { getEnumKeys } from './enum.js';
import { UnexpectedValueError } from './error.js';
import { indefinite } from './indefinite.js';
import { isNil, type Nillable } from './nil.js';
import type { Path } from './path.js';
import { isPlainObject, type PlainObject } from './plain-object.js';

export const isGraphQLResolveInfo = (
  maybeGraphQLResolveInfo: unknown,
): maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo =>
  isPlainObject(maybeGraphQLResolveInfo) &&
  typeof maybeGraphQLResolveInfo['fieldName'] === 'string' &&
  Array.isArray(maybeGraphQLResolveInfo['fieldNodes']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['returnType']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['parentType']);

export function assertGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
  path?: Path,
): asserts maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  if (!isGraphQLResolveInfo(maybeGraphQLResolveInfo)) {
    throw new UnexpectedValueError(
      `a GraphQLResolveInfo`,
      maybeGraphQLResolveInfo,
      { path },
    );
  }
}

export const isGraphQLASTNode = <TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
): maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] =>
  isPlainObject(maybeGraphQLASTNode) && maybeGraphQLASTNode['kind'] === kind;

export function assertGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
  path?: Path,
): asserts maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  if (!isGraphQLASTNode(maybeGraphQLASTNode, kind)) {
    throw new UnexpectedValueError(`a GraphQL ${kind}`, maybeGraphQLASTNode, {
      path,
    });
  }
}

export const parseGraphQLScalarValue = <TInternal>(
  type: graphql.GraphQLScalarType<TInternal, any>,
  maybeScalarValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeScalarValue)) {
    return maybeScalarValue;
  }

  try {
    return type.parseValue(maybeScalarValue);
  } catch (error) {
    throw new UnexpectedValueError(indefinite(type.name), maybeScalarValue, {
      cause: error,
      path,
    });
  }
};

export const parseGraphQLEnumValue = <TInternal = any>(
  type: graphql.GraphQLEnumType,
  maybeEnumValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeEnumValue)) {
    return maybeEnumValue;
  }

  const enumValue = type
    .getValues()
    .find(({ value }) => value === maybeEnumValue);

  if (!enumValue) {
    throw new UnexpectedValueError(
      `${indefinite(type.name)} (= a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}")`,
      maybeEnumValue,
      { path },
    );
  }

  return enumValue.value;
};

export const parseGraphQLLeafValue = (
  type: graphql.GraphQLLeafType,
  maybeLeafValue: unknown,
  path?: Path,
): any => {
  if (isNil(maybeLeafValue)) {
    return maybeLeafValue;
  }

  return graphql.isScalarType(type)
    ? parseGraphQLScalarValue(type, maybeLeafValue, path)
    : parseGraphQLEnumValue(type, maybeLeafValue, path);
};

export type CreateGraphQLEnumTypeOptions = {
  description?: Nillable<string>;
  useKeyAsValue?: OptionalFlag;
};

export const createGraphQLEnumType = (
  name: string,
  enumerable: PlainObject,
  options?: CreateGraphQLEnumTypeOptions,
): graphql.GraphQLEnumType =>
  new graphql.GraphQLEnumType({
    name,
    description: options?.description,
    values: Object.fromEntries(
      getEnumKeys(enumerable).map<[string, graphql.GraphQLEnumValueConfig]>(
        (key) => [
          key,
          {
            value: getOptionalFlag(options?.useKeyAsValue, false)
              ? undefined
              : enumerable[key],
          },
        ],
      ),
    ),
  });
