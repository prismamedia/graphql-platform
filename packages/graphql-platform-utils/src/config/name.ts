import * as graphql from 'graphql';
import { UnexpectedValueError } from '../error.js';
import type { Path } from '../path.js';

/**
 * A string compliant with the GraphQL specification
 *
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export type Name = string;

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function assertName(
  maybeName: unknown,
  path: Path,
): asserts maybeName is Name {
  if (typeof maybeName !== 'string' || !maybeName) {
    throw new UnexpectedValueError(`a non-empty string`, maybeName, {
      path,
    });
  }

  try {
    graphql.assertName(maybeName);
  } catch (error) {
    throw new UnexpectedValueError(
      'to be valid against the GraphQL "Names" specification (@see: https://spec.graphql.org/draft/#sec-Names)',
      maybeName,
      { cause: error, path },
    );
  }
}

/**
 * @see https://spec.graphql.org/draft/#sec-Names
 */
export function ensureName(maybeName: unknown, path: Path): Name {
  assertName(maybeName, path);

  return maybeName;
}
