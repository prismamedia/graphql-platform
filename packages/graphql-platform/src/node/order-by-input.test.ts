import { GraphQLEnumType, printType } from 'graphql';
import { GraphQLPlatform } from '..';
import { MyGP, nodeNames, nodes } from '../__tests__/config';
import {
  OrderByInput,
  TOrderByInputValue,
  TParsedOrderByInputValue,
} from './order-by-input';

describe('OrderByInput', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  it.each(nodeNames)('has a(n) %sOrderByInput', (nodeName) => {
    const node = gp.getNode(nodeName);
    const orderByInput = node.orderByInput;

    expect(orderByInput).toBeInstanceOf(OrderByInput);
    expect(orderByInput.public).toEqual(node.public);

    if (orderByInput.public) {
      if (orderByInput.type) {
        expect(orderByInput.type).toBeInstanceOf(GraphQLEnumType);
        expect(
          printType(orderByInput.type, { commentDescriptions: true }),
        ).toMatchSnapshot(orderByInput.name);
      }
    } else {
      expect(() => orderByInput.type).toThrowError(
        `"${nodeName}OrderByInput" is private`,
      );
    }
  });

  it.each([
    [
      'Article',
      undefined,
      `An error has occurred at ArticleOrderByInput - expects an array, got: undefined`,
    ],
    [
      'Article',
      null,
      `An error has occurred at ArticleOrderByInput - expects an array, got: null`,
    ],
    [
      'Article',
      ['_id_DESC', '_id_DESCS'],
      'An error has occurred at ArticleOrderByInput.1 - expects not to contain the unknown ordering expression "_id_DESCS", got: ["_id_DESC","_id_DESCS"]',
    ],
  ])(
    'throws an Error on %sOrderByInput.parseValue(%p)',
    (nodeName, value, error) => {
      const node = gp.getNode(nodeName);
      const orderByInput = node.orderByInput;

      expect(
        // @ts-expect-error
        () => orderByInput.parseValue(value),
      ).toThrowError(error);
    },
  );

  it.each<[string, TOrderByInputValue, TParsedOrderByInputValue]>([
    ['Article', [], []],
    [
      'Article',
      ['_id_DESC', 'createdAt_ASC'],
      [
        { kind: 'Leaf', leaf: '_id', direction: 'DESC' },
        { kind: 'Leaf', leaf: 'createdAt', direction: 'ASC' },
      ],
    ],
  ])('%sOrderByInput.parseValue(%p) = %p', (nodeName, value, parsedValue) => {
    const node = gp.getNode(nodeName);
    const orderByInput = node.orderByInput;

    expect(orderByInput.parseValue(value)).toEqual(parsedValue);
  });
});
