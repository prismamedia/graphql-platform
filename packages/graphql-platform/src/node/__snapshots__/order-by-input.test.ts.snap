// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderByInput has a(n) ArticleOrderByInput: ArticleOrderByInput 1`] = `
"# Used to sort the \\"Article\\" nodes
enum ArticleOrderByInput {
  # Sort the \\"Article\\" nodes from the lowest \\"createdAt\\" leaf to the highest
  createdAt_ASC

  # Sort the \\"Article\\" nodes from the highest \\"createdAt\\" leaf to the lowest
  createdAt_DESC

  # Sort the \\"Article\\" nodes from the lowest \\"updatedAt\\" leaf to the highest
  updatedAt_ASC

  # Sort the \\"Article\\" nodes from the highest \\"updatedAt\\" leaf to the lowest
  updatedAt_DESC
}"
`;

exports[`OrderByInput has a(n) ArticleTagOrderByInput: ArticleTagOrderByInput 1`] = `
"# Used to sort the \\"ArticleTag\\" nodes
enum ArticleTagOrderByInput {
  # Sort the \\"ArticleTag\\" nodes from the lowest \\"order\\" leaf to the highest
  order_ASC

  # Sort the \\"ArticleTag\\" nodes from the highest \\"order\\" leaf to the lowest
  order_DESC
}"
`;

exports[`OrderByInput has a(n) CategoryOrderByInput: CategoryOrderByInput 1`] = `
"# Used to sort the \\"Category\\" nodes
enum CategoryOrderByInput {
  # Sort the \\"Category\\" nodes from the lowest \\"order\\" leaf to the highest
  order_ASC

  # Sort the \\"Category\\" nodes from the highest \\"order\\" leaf to the lowest
  order_DESC
}"
`;

exports[`OrderByInput has a(n) HitOrderByInput: HitOrderByInput 1`] = `
"# Used to sort the \\"Hit\\" nodes
enum HitOrderByInput {
  # Sort the \\"Hit\\" nodes from the lowest \\"_id\\" leaf to the highest
  _id_ASC

  # Sort the \\"Hit\\" nodes from the highest \\"_id\\" leaf to the lowest
  _id_DESC

  # Sort the \\"Hit\\" nodes from the lowest \\"at\\" leaf to the highest
  at_ASC

  # Sort the \\"Hit\\" nodes from the highest \\"at\\" leaf to the lowest
  at_DESC
}"
`;

exports[`OrderByInput has a(n) TagOrderByInput: TagOrderByInput 1`] = `
"# Used to sort the \\"Tag\\" nodes
enum TagOrderByInput {
  # Sort the \\"Tag\\" nodes from the lowest \\"createdAt\\" leaf to the highest
  createdAt_ASC

  # Sort the \\"Tag\\" nodes from the highest \\"createdAt\\" leaf to the lowest
  createdAt_DESC

  # Sort the \\"Tag\\" nodes from the lowest \\"updatedAt\\" leaf to the highest
  updatedAt_ASC

  # Sort the \\"Tag\\" nodes from the highest \\"updatedAt\\" leaf to the lowest
  updatedAt_DESC
}"
`;

exports[`OrderByInput has a(n) UserProfileOrderByInput: UserProfileOrderByInput 1`] = `
"# Used to sort the \\"UserProfile\\" nodes
enum UserProfileOrderByInput {
  # Sort the \\"UserProfile\\" nodes from the lowest \\"birthday\\" leaf to the highest
  birthday_ASC

  # Sort the \\"UserProfile\\" nodes from the highest \\"birthday\\" leaf to the lowest
  birthday_DESC
}"
`;
